python_modules/dagster/dagster/_core/execution/run_metrics_thread.py

```diff
@@ -0,0 +1,296 @@
+        "location_name": location_name,
+        "run_id": dagster_run.run_id,
+    }
+    # filter out none values
+    return {k: v for k, v in tags.items() if v is not None}
+
+
+def _get_container_metrics(
+    logger: Optional[logging.Logger] = None,
+) -> Dict[str, Union[float, None]]:
+    metrics = retrieve_containerized_utilization_metrics(logger=logger)
+
+    # calculate cpu_limit
+    cpu_quota_us = metrics.get("cpu_cfs_quota_us")
+    cpu_period_us = metrics.get("cpu_cfs_period_us")
+    cpu_usage_ms = metrics.get("cpu_usage")
+    cpu_limit_ms = None
+    if cpu_quota_us and cpu_quota_us > 0 and cpu_period_us and cpu_period_us > 0:
+        # Why the 1000 factor is a bit counterintuitive:
+        # quota / period -> fraction of cpu per unit of time
+        # 1000 * quota / period -> ms/sec of cpu
+        cpu_limit_ms = (1000.0 * cpu_quota_us) / cpu_period_us
+
+    cpu_percent = None
+    if cpu_limit_ms and cpu_limit_ms > 0 and cpu_usage_ms and cpu_usage_ms > 0:
+        cpu_percent = 100.0 * cpu_usage_ms / cpu_limit_ms
+
+    memory_percent = None
+    memory_limit = metrics.get("memory_limit")
+    memory_usage = metrics.get("memory_usage")
+    if memory_limit and memory_limit > 0 and memory_usage and memory_usage > 0:
+        memory_percent = 100.0 * memory_usage / memory_limit
+
+    return {
+        "container.cpu_usage_ms": cpu_usage_ms,
+        "container.cpu_cfs_period_us": cpu_period_us,
+        "container.cpu_cfs_quota_us": cpu_quota_us,
+        "container.cpu_limit_ms": cpu_limit_ms,
+        "container.cpu_percent": cpu_percent,
+        "container.memory_usage": memory_usage,
+        "container.memory_limit": memory_limit,
+        "container.memory_percent": memory_percent,
+    }
+
+
+def _get_python_runtime_metrics() -> Dict[str, float]:
+    gc_stats = gc.get_stats()
+
+    stats_dict = {}
+    for index, gen_dict in enumerate(gc_stats):
+        gen_metrics = {
+            f"python.runtime.gc_gen_{index}.{key}": value for key, value in gen_dict.items()
+        }
+        stats_dict.update(gen_metrics)
+
+    return {**stats_dict, "python.runtime.gc_freeze_count": gc.get_freeze_count()}
+
+
+def _report_run_metrics_graphql(
+    instance: DagsterInstance,
+    dagster_run: DagsterRun,
+    metrics: Dict[str, float],
+    run_tags: Dict[str, str],
+):
+    datapoints: List[TelemetryDataPoint] = []
+    for metric, value in metrics.items():
+        if value is None:
+            continue
+        try:
+            datapoint = TelemetryDataPoint(
+                name=metric,
+                value=float(value),
+            )
+            datapoints.append(datapoint)
+        except ValueError:
+            logging.warning(f"Failed to convert metric value to float: {metric}={value}, skipping")
+
+    telemetry_data = RunTelemetryData(run_id=dagster_run.run_id, datapoints=datapoints)
+
+    instance._run_storage.add_run_telemetry(  # noqa: SLF001
+        telemetry_data, tags=run_tags
+    )
```

Comment:
am i reading this right that our write rate is the same as our polling rate? seems likely we want those to vary independently

Reply:
i am not sure why it would be worth it. 
- this is not high value, we don't care to cache or have any guarantee the data is sent 
- this is not high frequency data (polling should be every 30s or 60s, tbd) that would need bucketing.
- if we do bucket data to send in larger batch, we are likely to blind the last few value points of data that might tell us how a run was trending before getting killed.

Reply:
ah i see, I was thinking about supporting higher sampling rates. Agree that if we keep it low fidelity this is not as consequential